streams/src/main/java/org/apache/kafka/streams/kstream/internals/TransformerSupplierAdapter.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream.internals;
+
+import java.util.Arrays;
+
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.kstream.Transformer;
+import org.apache.kafka.streams.kstream.TransformerSupplier;
+import org.apache.kafka.streams.processor.ProcessorContext;
+
+public class TransformerSupplierAdapter<KIn, VIn, KOut, VOut> implements TransformerSupplier<KIn, VIn, Iterable<KeyValue<KOut, VOut>>> {
+
+    private TransformerSupplier<KIn, VIn, KeyValue<KOut, VOut>> transformerSupplier;
+
+    public TransformerSupplierAdapter(final TransformerSupplier<KIn, VIn, KeyValue<KOut, VOut>> transformerSupplier) {
+        this.transformerSupplier = transformerSupplier;
+    }
+
+    @Override
+    public Transformer<KIn, VIn, Iterable<KeyValue<KOut, VOut>>> get() {
+        return new Transformer<KIn, VIn, Iterable<KeyValue<KOut, VOut>>>() {
+
+            private Transformer<KIn, VIn, KeyValue<KOut, VOut>> transformer = transformerSupplier.get();
+
+            @Override
+            public void init(final ProcessorContext context) {
+                transformer.init(context);
+            }
+
+            @Override
+            public Iterable<KeyValue<KOut, VOut>> transform(final KIn key, final VIn value) {
+                final KeyValue<KOut, VOut> pair = transformer.transform(key, value);
+                if (pair != null) {
+                    return Arrays.asList(pair);
```

Comment:
nit: Collections.singletonList.

Reply:
@guozhangwang , you had a performance concern about this class.

It definitely pads the execution, but my gut says that it won't be too bad, and the complexity savings of consolidating the code paths probably makes up for it.

`Collections.singletonList` doesn't create a full-blown list, but just wraps the argument in an instance of `SingletonList`, so it's very light to allocate and _very_ light to collect. `Collections.emptyList` just returns a reference to a static singleton instance of `EmptyList`, so there's no overhead to it.

Also, the iteration should be low/no cost as well in hot code, as the jit compiler should unroll the loop.

This is all speculative, but something that @cadonna could evaluate with a microbenchmark if desired.

Anyway, like I said, I'm not worried about it in this case.

Reply:
oh. I do recommend taking both your suggestions to use `singletonList`/`emptyList` in this class, and actually in general as well.

Reply:
I replaced Arrays.asList with Collections.singletonList.

Reply:
Thanks! (FYI, if you use IDEA, there's a code inspection you can enable to prompt you to make this change.)

Reply:
Cool, I think I'm convinced. If others like `TransformerSupplierAdapter` for complexity savings as well we can merge this as-is.

Reply:
LGTM. @mjsax please make a final pass and merge if you have no further comments.